### PR TITLE
Maintainer updating name service for load balancers 

### PR DESCRIPTION
**Pull request by tokle**

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java [hunk 1]
```diff
@@ -147,7 +147,7 @@ public LockedApplication withNewDeployment(ZoneId zone, ApplicationVersion appli
                                                   previousDeployment.clusterUtils(),
                                                   previousDeployment.clusterInfo(),
                                                   previousDeployment.metrics(),
-                                                  previousDeployment.activity());
+                                                  previousDeployment.activity(), Collections.emptyMap());
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java]: This should be `previousDeployment.loadBalancers()` so load balancers are not thrown away when updating deployments.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/Deployment.java [hunk 2]
```diff
@@ -46,6 +48,7 @@ public Deployment(ZoneId zone, ApplicationVersion applicationVersion, Version ve
         this.clusterInfo = Objects.requireNonNull(clusterInfo, "clusterInfo cannot be null");
         this.metrics = Objects.requireNonNull(metrics, "deploymentMetrics cannot be null");
         this.activity = Objects.requireNonNull(activity, "activity cannot be null");
+        this.loadBalancers = Objects.requireNonNull(loadBalancers, "loadBalancers cannot be null");
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/Deployment.java]: Please wrap this with `ImmutableMap#copyOf`. The same should be done for the `clusterUtilization` and `clusterInfo` fields.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/ControllerMaintenance.java [hunk 3]
```diff
@@ -81,6 +82,7 @@ public ControllerMaintenance(MaintainerConfig maintainerConfig, ApiAuthorityConf
         osVersionStatusUpdater = new OsVersionStatusUpdater(controller, maintenanceInterval, jobControl);
         contactInformationMaintainer = new ContactInformationMaintainer(controller, Duration.ofHours(12), jobControl, contactRetriever);
         costReportMaintainer = new CostReportMaintainer(controller, Duration.ofHours(2), reportConsumer, jobControl, nodeRepositoryClient, Clock.systemUTC(), selfHostedCostConfig);
+        loadbalancerMaintainer = new LoadBalancerMaintainer(controller, Duration.ofMinutes(5), jobControl, nameService);
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/ControllerMaintenance.java]: This should be deconstructed in the `deconstruct` method.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        List<Application> applications = controller().applications().asList();
+        applications.forEach(this::registerLoadBalancerEndpoint);
+
+        // Delete removed application rotations
+        // TODO
+    }
+
+    private void updateApplicationLoadBalancers(Application application) {
+        // Get a list of all load balancers for this applications (for all zones and clusters)
+        Map<ZoneId, List<LoadBalancer>> applicationZoneLoadBalancers = new HashMap<>();
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java]: Consider just `zoneLoadBalancers`, application dimension is implied by method name.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        for (ZoneId zoneId : application.deployments().keySet()) {
+            try {
+                applicationZoneLoadBalancers.put(zoneId, controller().applications().configServer().getLoadBalancers(new DeploymentId(application.id(), zoneId)));
+            } catch (Exception e) {
+                log.log(LogLevel.INFO,
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java]: Warning? Also consider adding `"Retrying in " + maintenanceInterval()` to the message.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/configserver/LoadBalancer.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import com.yahoo.vespa.hosted.controller.api.identifiers.InstanceId;
+import com.yahoo.vespa.hosted.controller.api.identifiers.TenantId;
+
+public class LoadBalancer {
+    private final String id;
+    private final TenantId tenant;
+    private final ApplicationId application;
+    private final InstanceId instance;
+    private final String cluster;
+    private final String hostname;
```
Comment (mpolden) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/configserver/LoadBalancer.java]: Type should be `com.yahoo.config.provision.HostName`.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/configserver/LoadBalancer.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import com.yahoo.vespa.hosted.controller.api.identifiers.ApplicationId;
+import com.yahoo.vespa.hosted.controller.api.identifiers.InstanceId;
+import com.yahoo.vespa.hosted.controller.api.identifiers.TenantId;
+
+public class LoadBalancer {
+    private final String id;
+    private final TenantId tenant;
+    private final ApplicationId application;
+    private final InstanceId instance;
+    private final String cluster;
```
Comment (mpolden) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/configserver/LoadBalancer.java]: Type should be `com.yahoo.config.provision.ClusterSpec.Id`.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    private void registerLoadBalancerEndpoint(Application application) {
+        for (Map.Entry<ZoneId, Deployment> deploymentEntry : application.deployments().entrySet()) {
+            ZoneId zone = deploymentEntry.getKey();
+            Deployment deployment = deploymentEntry.getValue();
+            for (Map.Entry<ClusterSpec.Id, HostName> loadBalancers : deployment.loadBalancers().entrySet()) {
+                try {
+                    RecordName recordName = RecordName.from(getEndpointName(loadBalancers.getKey(), application.id(), zone));
+                    RecordData recordData = RecordData.fqdn(loadBalancers.getValue().value());
+                    nameService.createCname(recordName, recordData);
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java]: As discussed offline: This should use a combination of `NameService#findRecord` and not assume that implementation of `NameService#createCname` is idempotent.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+                } catch (Exception e) {
+                    // Catching any exception, will be retried on next run
+                    log.log(LogLevel.INFO,
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java]: Warning? Also consider adding `"Retrying in " + maintenanceInterval()` to the message.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializer.java [hunk 10]
```diff
@@ -499,6 +509,12 @@ private Change changeFromSlime(Inspector object) {
         return field.valid() ? optionalString(field).map(RotationId::new) : Optional.empty();
     }
 
+    private Map<ClusterSpec.Id, HostName> loadBalancerMapFromSlime(Inspector object) {
+        Map<ClusterSpec.Id, HostName> map = new HashMap<>();
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializer.java]: Consider a more descriptive name.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+public class LoadBalancerMaintainer extends Maintainer {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java]: Add javadoc with a short description of what this does and author.

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainerTest.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * @author mpolden
```
Comment (mpolden) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainerTest.java]: Is this for plausible deniability? 🙂

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/configserver/LoadBalancer.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2019 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.controller.api.integration.configserver;
+
+import com.yahoo.vespa.hosted.controller.api.identifiers.ApplicationId;
+import com.yahoo.vespa.hosted.controller.api.identifiers.InstanceId;
+import com.yahoo.vespa.hosted.controller.api.identifiers.TenantId;
+
+public class LoadBalancer {
```
Comment (mpolden) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/configserver/LoadBalancer.java]: Add javadoc with short description and author.

**Review by mpolden — CHANGES_REQUESTED** [review]
Good start, please consider my comments.

**Comment by tokle** [comment]
@andreer fyi

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/persistence/LoadBalancerNameSerializerTest.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+                ApplicationId.from("foo", "bar", "default"), Lists.newArrayList(new LoadBalancerName(new RecordId("123.4123.:124123"), RecordName.from("foo.bar"))));
+        Slime slime = serializer.toSlime(lbnames);
+        byte[] bytes = SlimeUtils.toJsonBytes(slime);
+        System.out.println("json = " + new String(bytes));
+
+        Map<ApplicationId, List<LoadBalancerName>> applicationIdListMap = serializer.fromSlime(SlimeUtils.jsonToSlime(bytes));
+        applicationIdListMap.forEach((k,v) -> {
+            System.out.println("k = " + k);
+            System.out.println("v = " + v);
+        });
```
Comment (mpolden) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/persistence/LoadBalancerNameSerializerTest.java]: Missing asserts?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/LoadBalancerNameSerializer.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    public Map<ApplicationId, List<LoadBalancerName>> fromSlime(Slime slime) {
+
+        Inspector object = slime.get().field(loadBalancerNamesField);
+        Map<ApplicationId, List<LoadBalancerName>> loadBalancerNames = new HashMap<>();
+        object.traverse((ObjectTraverser) (appId, inspector) ->
+                loadBalancerNames.put(ApplicationId.fromSerializedForm(appId), loadBalancerNamesFromSlime(inspector)));
+//
+//        Set<OsVersion> osVersions = new TreeSet<>();
+//        array.traverse((ArrayTraverser) (i, inspector) -> osVersions.add(fromSlime(inspector)));
+//        return Collections.unmodifiableSet(osVersions);
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/LoadBalancerNameSerializer.java]: Remove?

**Review by mpolden — APPROVED** [review]

**Pull request merged**